pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -354,6 +362,71 @@ impl WalreceiverState {
         });
     }
 
+    /// Drops the current connection (if any) and updates retry timeout for the next
+    /// connection attempt to the same safekeeper.
+    async fn drop_old_connection(&mut self, needs_shutdown: bool) {
+        let wal_connection = match self.wal_connection.take() {
+            Some(wal_connection) => wal_connection,
+            None => return,
+        };
+
+        if needs_shutdown {
+            wal_connection.connection_task.shutdown().await;
+        }
+
+        let retry = self
+            .wal_connection_retries
+            .entry(wal_connection.sk_id)
+            .or_insert(RetryInfo {
+                next_retry_at: None,
+                retry_duration_seconds: WALCONNECTION_RETRY_MIN_BACKOFF_SECONDS,
+            });
+
+        let now = Utc::now().naive_utc();
+
+        // Schedule the next retry attempt. We want to have exponential backoff for connection attempts,
+        // and we add backoff to the time when we started the connection attempt. If the connection
+        // was active for a long time, then next_retry_at will be in the past.
+        retry.next_retry_at =
+            wal_connection
+                .started_at
+                .checked_add_signed(chrono::Duration::milliseconds(
+                    (retry.retry_duration_seconds * 1000.0) as i64,
```

Comment:
Quite interesting, there's no `f64 <-> i64` conversion via [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) is implemented.

f64's integer part range seems narrow enough ok to cast into i64, but I wonder if I'm missing anything.

Reply:
So, do you know any better way to convert f64 seconds to chrono::Duration? I don't like multiplying to 1000.0 like it's done now, looks very hacky.

Reply:
I'm afraid, there's no simple float <-> integer conversion anywhere I've looked at, same for overflow checks, so not sure how to make it better conceptually.